File: world/examples/frozenlake_qtable.py
import numpy as np
import world

env = world.make("FrozenLake-v1")
print('env=', env)
print('env.observation_space=', env.observation_space)
print('env.action_space=', env.action_space)
print('env.metadata=', env.metadata)

# 超參數設定
alpha = 0.8          # 學習速率
gamma = 0.95         # 折扣因子
num_episodes = 2000  # 訓練回合數

# 初始化 Q 表：16 個狀態 × 4 個動作
Q = np.zeros([env.observation_space.n, env.action_space.n])

# 選擇演算法（取消註解切換）
# method = "Q"
method = "SARSA"
# method = "TD_LAMBDA"
if method == "TD_LAMBDA":
    lambda_ = 0.2           # 資格跡衰減率

for i in range(num_episodes):
    s, info = env.reset()

    if method == "TD_LAMBDA":
        # 每個回合初始化資格跡矩陣
        E = np.zeros([env.observation_space.n, env.action_space.n])

    for j in range(99):
        # ε-greedy 選擇動作：加入隨機噪聲實現探索，噪聲隨回合遞減
        a = np.argmax(Q[s,:] + np.random.randn(1, env.action_space.n) * (1./(i+1)))
        s1, reward, terminated, truncated, info = env.step(a)

        if method == "Q":
            # Q-Learning：使用 max_{a'} Q(s',a') 進行離策略更新
            Q[s,a] += alpha * (reward + gamma * np.max(Q[s1,:]) - Q[s,a])

        elif method == "SARSA":
            # SARSA：使用實際選擇的 a' 對應的 Q(s',a') 進行基於策略更新
            a1 = np.argmax(Q[s1, :] + np.random.randn(1, env.action_space.n) * (1. / (i + 1)))
            Q[s, a] += alpha * (reward + gamma * Q[s1, a1] - Q[s, a])

        elif method == "TD_LAMBDA":
            # TD(λ)：結合 TD 誤差與資格跡，實現更平滑的信用分配
            a1 = np.argmax(Q[s1, :] + np.random.randn(1, env.action_space.n) * (1. / (i + 1)))
            delta = reward + gamma * Q[s1, a1] - Q[s, a]  # TD 誤差

            # 全局衰減資格跡，再對當前 (s,a) 增量
            for s2 in range(env.observation_space.n):
                for a2 in range(env.action_space.n):
                    E[s2, a2] *= gamma * lambda_
            E[s,a] += 1

            Q[s, a] += alpha * delta * E[s, a]

        else:
            raise Exception(f"無法處理的 method={method}")

        s = s1
        if terminated == True:
            break

print('Q=', Q)

if method == "TD_LAMBDA":
    print('E=', E)

print('完成迭代，展示學習成果 ...')

# 使用訓練好的 Q 表進行視覺化展示（永遠取最佳動作）
env = world.make('FrozenLake-v1')
s, info = env.reset()
for i in range(100):
    env.render()
    a = np.argmax(Q[s,:])
    s1, reward, terminated, truncated, info = env.step(a)
    print('-'*30)
    s = s1
    if terminated == True:
        break

File: world/examples/frozenlake_qtable.ts
/**
 * examples/frozenlake_qtable.ts
 * Q-Learning, SARSA, and TD(λ) comparison for FrozenLake-v1
 */
import { make } from '../index';

const ALPHA = 0.8;
const GAMMA = 0.95;
const NUM_EPISODES = 2000;
const LAMBDA = 0.2;

const method = 'SARSA'; // 'Q', 'SARSA', or 'TD_LAMBDA'

function trainFrozenLake(): void {
  const env = make('FrozenLake-v1');
  const nStates = (env as any).observationSpace?.n ?? 16;
  const nActions = (env as any).actionSpace?.n ?? 4;

  const Q: number[][] = Array.from({ length: nStates }, () => new Array(nActions).fill(0));

  console.log('='.repeat(50));
  console.log(`  Training: ${method}`);
  console.log('='.repeat(50));

  for (let i = 0; i < NUM_EPISODES; i++) {
    const resetResult = env.reset();
    let s = resetResult.observation as number;

    let E: number[][] | null = null;
    if (method === 'TD_LAMBDA') {
      E = Array.from({ length: nStates }, () => new Array(nActions).fill(0));
    }

    for (let j = 0; j < 99; j++) {
      const noise = 1.0 / (i + 1);
      const action = argmaxWithNoise(Q[s], noise);

      const result = env.step(action);
      const s1 = result.observation as number;
      const reward = result.reward;
      const done = result.terminated || result.truncated;

      if (method === 'Q') {
        Q[s][action] += ALPHA * (reward + GAMMA * Math.max(...Q[s1]) - Q[s][action]);
      } else if (method === 'SARSA') {
        const a1 = argmaxWithNoise(Q[s1], noise);
        Q[s][action] += ALPHA * (reward + GAMMA * Q[s1][a1] - Q[s][action]);
      } else if (method === 'TD_LAMBDA') {
        const a1 = argmaxWithNoise(Q[s1], noise);
        const delta = reward + GAMMA * Q[s1][a1] - Q[s][action];

        for (let s2 = 0; s2 < nStates; s2++) {
          for (let a2 = 0; a2 < nActions; a2++) {
            E![s2][a2] *= GAMMA * LAMBDA;
          }
        }
        E![s][action] += 1;

        for (let s2 = 0; s2 < nStates; s2++) {
          for (let a2 = 0; a2 < nActions; a2++) {
            Q[s2][a2] += ALPHA * delta * E![s2][a2];
          }
        }
      }

      s = s1;
      if (done) break;
    }

    if ((i + 1) % 500 === 0) {
      const success = evaluate(env, Q);
      console.log(`  Episode ${i + 1}: success rate = ${(success * 100).toFixed(1)}%`);
    }
  }

  console.log('\n  Q table:');
  console.log(Q.map((row, i) => `  s${i}: [${row.map(v => v.toFixed(2)).join(', ')}]`).join('\n'));

  // Demo
  console.log('\n  Demonstrating learned policy ...');
  const demoEnv = make('FrozenLake-v1');
  const demoReset = demoEnv.reset();
  let s = demoReset.observation as number;
  for (let i = 0; i < 100; i++) {
    demoEnv.render();
    const a = Q[s].indexOf(Math.max(...Q[s]));
    const result = demoEnv.step(a);
    console.log('-'.repeat(30));
    s = result.observation as number;
    if (result.terminated) break;
  }
  demoEnv.close();
}

function argmaxWithNoise(Qs: number[], noise: number): number {
  const noisy = Qs.map(q => q + Math.random() * noise);
  let best = 0;
  for (let a = 1; a < noisy.length; a++) {
    if (noisy[a] > noisy[best]) best = a;
  }
  return best;
}

function evaluate(env: ReturnType<typeof make>, Q: number[][]): number {
  let wins = 0;
  for (let ep = 0; ep < 100; ep++) {
    const resetResult = env.reset({ seed: ep });
    let s = resetResult.observation as number;
    for (let step = 0; step < 99; step++) {
      const a = Q[s].indexOf(Math.max(...Q[s]));
      const result = env.step(a);
      s = result.observation as number;
      if (result.done) {
        if (result.reward > 0) wins++;
        break;
      }
    }
  }
  return wins / 100;
}

trainFrozenLake();